File: matlab/statictorquecalculation.m
%% Static Torque Calculation Tool (T = m * g * R)
% Author: Gil Soco
% Description:
%   Computes static torque for various masses and lever lengths using:
%               T = m * g * R
%   Includes:
%       - Automatic table generation
%       - Engineering plots
%
%   Outputs:
%       static_torque_results.csv
%       static_torque_plot.png

g = 9.81; % Acceleration due to Gravity (m/s^2)
masses = [3, 7, 10]; % Masses (m) in kg
lever_lengths = 0.1:0.1:1.0; % Lever Lengths (R) in meters (0.1 to 1.0)


num_rows = length(masses) * length(lever_lengths);
Mass = zeros(num_rows, 1);
LeverLength = zeros(num_rows, 1);
StaticTorque = zeros(num_rows, 1);
idx = 1;


scriptFolder = fileparts(mfilename('fullpath'));
projectFolder = fileparts(scriptFolder);
outputFolder = fullfile(projectFolder, 'Images');
if ~exist(outputFolder, 'dir')
    mkdir(outputFolder);
end

% Calculate Torque for every combination
disp('Calculating Static Torque (T = m * g * R)...');
disp('------------------------------------------');

for m = masses
    for R = lever_lengths
        % Static Torque formula
        T = m * g * R;
        
        % Store results
        Mass(idx) = m;
        LeverLength(idx) = R;
        StaticTorque(idx) = T;
        
        idx = idx + 1;
    end
end


TorqueTable = table(Mass, LeverLength, StaticTorque);

TorqueTable.Properties.VariableNames = {'Mass_kg', 'LeverLength_m', 'StaticTorque_Nm'};

disp(' ');
disp('Calculated Static Torque Results');
disp(TorqueTable);


writetable(TorqueTable, 'm-static_torque_results.csv');
disp(' ');
disp('Results also saved to m-static_torque_results.csv');

%% Plot: Torque vs Lever Length for Each Mass
figure;
hold on; grid on;

unique_masses = unique(Mass);

for m = unique_masses'
    idx = (Mass == m);
    plot(LeverLength(idx), StaticTorque(idx), '-o', 'LineWidth', 1.5, ...
        'DisplayName', sprintf('Mass = %d kg', m));
end

xlabel('Lever Length (m)');
ylabel('Static Torque (N·m)');
title('Torque vs Lever Length for Different Masses');
legend('Location', 'northwest');
hold off;

%% Saves the plot as an image .png file
outputName_png = fullfile(outputFolder, 'm-static_torque_plot.png');

exportgraphics(gcf, outputName_png, 'Resolution', 300);   % High-quality PNG

disp(['Plot saved to folder: ' outputFolder]);

File: python/statictorquecalculation.py
##
#   Author: Gil Soco
#   Description:
#    This module provides functions to calculate static torque for various mechanical systems.
#                           T = m * g * R
#   Includes:
#       - Automatic table ggeneration
#       - Engineering plots
#
#   Outputs:
#       py-static_torque_results.csv
#       py-static_torque_plot.png

import matplotlib.pyplot as plt

g = 9.81  # Acceleration due to Gravity (m/s^2)
masses = [3, 7, 10]  # Masses (m) in kg
lever_lengths = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1]  # Lever lengths (R) in meters

def calculate_static_torque(mass, lever_length):
    """
    Calculate static torque.
    
    Parameters:
    mass (float): Mass in kg.
    lever_length (float): Lever length in meters.
    
    Returns:
    float: Static torque in Nm.
    """
    return mass * g * lever_length  

def generate_torque_table(masses, lever_lengths):
    """
    Generate a table of static torque values for given masses and lever lengths.
    
    Parameters:
    masses (list): List of masses in kg.
    lever_lengths (list): List of lever lengths in meters.
    
    Returns:
    list: List of tuples containing (mass, lever_length, torque).
    """
    torque_table = []
    for mass in masses:
        for lever_length in lever_lengths:
            torque = calculate_static_torque(mass, lever_length)
            torque_table.append((mass, lever_length, torque))
    return torque_table

def save_torque_table_to_csv(torque_table, filename='CSV/py-static_torque_results.csv'):
    """
    Save the torque table to a CSV file.
    
    Parameters:
    torque_table (list): List of tuples containing (mass, lever_length, torque).
    filename (str): Name of the output CSV file.
    """
    import csv
    with open(filename, mode='w', newline='') as file:
        writer = csv.writer(file)
        writer.writerow(['Mass (kg)', 'Lever Length (m)', 'Static Torque (Nm)'])
        for row in torque_table:
            writer.writerow(row)

def plot_torque_values(torque_table, filename='py-static_torque_plot.png'):
    """
    Plot static torque values and save the plot as an image.
    
    Parameters:
    torque_table (list): List of tuples containing (mass, lever_length, torque).
    filename (str): Name of the output image file.
    """
    
    masses = [row[0] for row in torque_table]
    lever_lengths = [row[1] for row in torque_table]
    torques = [row[2] for row in torque_table]
    
    plt.figure(figsize=(10, 6))
    scatter = plt.scatter(lever_lengths, torques, c=masses, cmap='viridis', s=100)
    plt.colorbar(scatter, label='Mass (kg)')
    plt.title('Static Torque vs Lever Length')
    plt.xlabel('Lever Length (m)')
    plt.ylabel('Static Torque (Nm)')
    plt.grid(True)
    plt.savefig("Images/"+filename)
    plt.close()

if __name__ == "__main__":
    torque_table = generate_torque_table(masses, lever_lengths)
    save_torque_table_to_csv(torque_table)
    plot_torque_values(torque_table)